### PR TITLE
MGMT-4893: Support must gather images in kube-api

### PR DESCRIPTION
**Pull request by slaviered**
# Assisted Pull Request

## List all the issues related to this PR

- [x] New Feature
- [ ] Bug fix
- [ ] Tests
- [ ] Documentation
- [ ] CI/CD 

## What environments does this code impact?

- [ ] Automation (CI, tools, etc)
- [ ] Cloud
- [x] Operator Managed Deployments
- [ ] None

## How was this code tested?

- [ ] assisted-test-infra environment
- [ ] dev-scripts environment
- [ ] Reviewer's test appreciated
- [x] Waiting for CI to do a full test run
- [ ] Manual (Elaborate on how it was tested)
- [ ] No tests needed

## Assignees

/cc @
/cc @

## Checklist

- [ ] Title and description added to both, commit and PR.
- [ ] Relevant issues have been associated (see [CONTRIBUTING] guide)
- [ ] Reviewers have been listed
- [ ] This change does not require a documentation update (docstring, `docs`, README, etc)
- [ ] Does this change include unit-tests (note that code changes require unit-tests)

## Reviewers Checklist

- [ ] Are the title and description (in both PR and commit) meaningful and clear?
- [ ] Is there a bug required (and linked) for this change?
- [ ] Should this PR be backported?

[Kubernetes community documentation]: https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#commit-message-guidelines
[CONTRIBUTING]: https://github.com/openshift/assisted-service/blob/master/CONTRIBUTING.md

### CODE REVIEW

File: internal/controller/controllers/agentserviceconfig_controller.go [hunk 1]
```diff
@@ -1024,30 +1025,57 @@ func (r *AgentServiceConfigReconciler) newAssistedServiceDeployment(ctx context.
...
+		mustGatherVersions[versionKey][specImage.Name] = specImage.Url
+	}
+
+	if len(mustGatherVersions) == 0 {
+		return MustGatherImages()
+	}
+	encodedVersions, err := json.Marshal(mustGatherVersions)
+	if err != nil {
+		log.Error(err, fmt.Sprintf("Problem marshaling must gather images (%v) to string, returning default %v", mustGatherVersions, MustGatherImages()))
+		return OpenshiftVersions()
```
Comment (carbonin) [inline internal/controller/controllers/agentserviceconfig_controller.go:1051]: Is this a typo? This should probably be `MustGatherImages()`, right?

Comment (slaviered) [reply internal/controller/controllers/agentserviceconfig_controller.go:1051]: Good Catch. Fixing

**Review by djzager — COMMENTED** [review]
First pass looks great! 

If you add the "default" upstream values in https://github.com/openshift/assisted-service/blob/master/config/samples/agent-install.openshift.io_v1beta1_agentserviceconfig.yaml

you can run `make generate-manifests` and it should update the clusterserviceversion with the sample agentserviceconfig. This would make it so when people use the UI to install the operator, these fields would be auto-populated.

File: internal/controller/controllers/agentserviceconfig_controller.go [hunk 2]
```diff
@@ -1024,30 +1025,57 @@ func (r *AgentServiceConfigReconciler) newAssistedServiceDeployment(ctx context.
 	return deployment, mutateFn, nil
 }
 
+func (r *AgentServiceConfigReconciler) getMustGatherImages(log logrus.FieldLogger, instance *aiv1beta1.AgentServiceConfig) string {
```
Comment (flaper87) [inline internal/controller/controllers/agentserviceconfig_controller.go:1028]: Would it be possible to add some docstrings on what this function does, the expected input, and the expected output?

**Comment by slaviered** [comment]
The "default" values will be added in a separate PR

**Review by flaper87 — CHANGES_REQUESTED** [review]
just a small request, otherwise lgtm.

In addition to the in-line docs, could you update [these docs](https://github.com/openshift/assisted-service/blob/master/docs/operator.md) to document this attribute in the spec?

I would love all of us to request more documentation (not only in-code but also user-facing docs)

File: internal/controller/controllers/agentserviceconfig_controller.go (hunk 2 above)
Comment (slaviered) [reply internal/controller/controllers/agentserviceconfig_controller.go:1028]: Sure. I will add this and the extra documentation in the following PR.

**Comment by slaviered** [comment]
Summary:
For the following PR the following should be added:
1. Adding default images values (for 4.8) to  https://github.com/openshift/assisted-service/blob/master/config/samples/agent-install.openshift.io_v1beta1_agentserviceconfig.yaml
2. Add comment to getMustGatherImages()
3. Add documentation to operator.md

**Review by flaper87 — COMMENTED** [review]
FWIW, it's fine to add these docs in the following-pr. I generally prefer docs to be written with the feature itself, but this PR is ready to go.

**Review by djzager — APPROVED** [review]
/lgtm

In the follow-up PR the things I would like to see are:
1. Documentation of when/how to use this in https://github.com/openshift/assisted-service/blob/master/docs/operator.md
1. Docstrings on some of these private methods that explain what they do and what their output is
1. A functional example added to https://github.com/openshift/assisted-service/blob/master/config/samples/agent-install.openshift.io_v1beta1_agentserviceconfig.yaml

**Pull request merged**